crates/rpc/rpc/src/otterscan.rs

```diff
@@ -205,7 +211,69 @@ where
     }
 
     /// Handler for `getContractCreator`
-    async fn get_contract_creator(&self, _address: Address) -> RpcResult<Option<ContractCreator>> {
-        Err(internal_rpc_err("unimplemented"))
+    async fn get_contract_creator(&self, address: Address) -> RpcResult<Option<ContractCreator>> {
+        if !self.has_code(address, None).await? {
+            return Ok(None);
+        }
+
+        // use binary search from block [1, latest block number] to find the first block where the
+        // contract was deployed
+        let mut low = 1;
+        let mut high = self.eth.block_number()?.try_into().unwrap();
+        let mut num = None;
+
+        while low < high {
+            let mid = (low + high) / 2;
+            let block_number = Some(BlockId::Number(BlockNumberOrTag::Number(mid)));
+            if self.eth.get_code(address, block_number).await?.is_empty() {
+                low = mid + 1; // not found in current block, need to search in the later blocks
+            } else {
+                high = mid - 1; // found in current block, try to find a lower block
+                num = Some(mid);
+            }
+        }
+
+        // this should not happen, only if the state of the chain is inconsistent
+        if num.is_none() {
+            return Ok(None);
+        }
+
+        let traces = self
+            .eth
+            .trace_block_with(
+                num.unwrap().into(),
+                TracingInspectorConfig::default_parity(),
+                |tx_info, inspector, _, _, _| {
+                    Ok(inspector.into_parity_builder().into_localized_transaction_traces(tx_info))
+                },
+            )
+            .await?
+            .map(|traces| {
+                traces
+                    .into_iter()
+                    .flatten()
+                    .filter_map(|tx_trace| {
+                        let trace = tx_trace.trace;
+                        match (trace.action, trace.result, trace.error) {
+                            (
+                                Action::Create(CreateAction { from: creator, .. }),
+                                Some(TraceOutput::Create(CreateOutput {
+                                    address: contract, ..
+                                })),
+                                None,
+                            ) if contract == address => Some(ContractCreator {
+                                hash: tx_trace.transaction_hash.unwrap_or_default(),
```

Comment:
I think we should rather do an `ok_or` here, because all `TransactionInfo` returned from tracing should have a non-empty `hash` field

Reply:
I'm sorry, could you please give more advice, the txhash here must not be none, so do you means if it is none, we should return an error(eg: `internal_rpc_err("txhash not found")`) or anything else?, if so, I'm not sure how to return the error in the filter_map closure, seems `?` not working.

Reply:
> so do you means if it is none, we should return an error(eg: internal_rpc_err("txhash not found")) 

yeah, it should be an error, because otherwise we'll return `0x0` as a transaction hash which is not correct

> if so, I'm not sure how to return the error in the filter_map closure

```diff
diff --git a/crates/rpc/rpc/src/otterscan.rs b/crates/rpc/rpc/src/otterscan.rs
index c60e8a6b3..e5d40f6ac 100644
--- a/crates/rpc/rpc/src/otterscan.rs
+++ b/crates/rpc/rpc/src/otterscan.rs
@@ -254,9 +254,9 @@ where
                 traces
                     .into_iter()
                     .flatten()
-                    .filter_map(|tx_trace| {
+                    .map(|tx_trace| {
                         let trace = tx_trace.trace;
-                        match (trace.action, trace.result, trace.error) {
+                        Ok(match (trace.action, trace.result, trace.error) {
                             (
                                 Action::Create(CreateAction { from: creator, .. }),
                                 Some(TraceOutput::Create(CreateOutput {
@@ -264,14 +264,18 @@ where
                                 })),
                                 None,
                             ) if contract == address => Some(ContractCreator {
-                                hash: tx_trace.transaction_hash.unwrap_or_default(),
+                                hash: tx_trace.transaction_hash.ok_or_else(|| {
+                                    internal_rpc_err("transaction hash is missing")
+                                })?,
                                 creator,
                             }),
                             _ => None,
-                        }
+                        })
                     })
-                    .collect::<Vec<_>>()
-            });
+                    .filter_map(Result::transpose)
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .transpose()?;
 
         // A contract maybe created and then destroyed in multiple transactions, here we
         // return the first found transaction, this behavior is consistent with etherscan's

```

Reply:
it's really helpful, thanks